### ISSUE TITLE
Typescript output?

### ISSUE THREAD

**Issue by StevenLangbroek** [salesforce-ux/theo#174]
I see there's no option for outputting typescript, but formats look relatively straightforward to write. Would you be interested in a typescript formatter?

Thanks!

- Steven

**Comment by trazek** [comment]
I'd be interested in seeing what that looks like.

**Comment by StevenLangbroek** [comment]
> I'd be interested in seeing what that looks like.

Initially probably not that different from the common.js one. I'm still exploring Theo though, so depending on how it works it might be interesting to output an `enum` or `union` per category.

**Comment by aputinski** [comment]
Yeah, I think Typescript could be interesting if we can output a format that is useful for everyone.

**Comment by petekp** [comment]
Our design system is Typescript-based and I wanted to leverage autocomplete in VSCode when referencing Theo-generated styles in a component. The quickest solution was to auto-generate typings for the JS module using `tsc` and its `--declaration` flag. Example: https://github.com/VirtaHealth/substrate/blob/master/packages/styles/dist/index.d.ts

It only generates the primitive types, e.g. `string`, `number`, etc. which has been sufficient for autocompleting theme properties. These annotations could be further refined to specific tokens, but I haven't found this necessary just yet.

**Comment by trazek** [comment]
@petekp that link gave me a 404.

**Comment by petekp** [comment]
@trazek Oops, fixed :)

**Comment by mattfelten** [comment]
@petekp Could you walk through how you generated the typings for that? `tsc --declaration tokens.js` doesn't love that it's not a TS file.

**Comment by petekp** [comment]
@mattfelten I'm currently faking it by registering a custom format (`.d.ts`) and outputting the style data but wrapped in a typings export declaration; basically producing the same output `tsc --declaration` would emit:

```js
// Sort style properties 
const categorizeStyles = (styleMap: ImmutableStyleMap) => {
  const styleProps: ThemeStyle[] = styleMap.get('props').toJS()
  return chain(styleProps)
    .sortBy('name')
    .groupBy('category')
    .value()
}

export const createThemeObjectFromStyles = (
  styleData: ImmutableStyleMap
): Readonly<Theme> => {
  return Object.entries(categorizeStyles(styleData)).reduce(
    (acc, [category, styles]) => ({
      ...acc,
      ...styles.reduce((prev, style) => {
        return {
          ...prev,
          [style.name]: style.value,
        }
      }, {}),
    }),
    {}
  )
}


theo.registerFormat(
    'd.ts',
    styleMap =>
      `export declare const _default: 
        ${JSON.stringify(createThemeObjectFromStyles(styleMap))}
      `
  )
```

**Comment by mattfelten** [comment]
Ah that's clever. I ended up registering a new `ts` format and then running `tsc` on it.

```
module.exports = theo => {
	theo.registerFormat(
		"ts",
`
// Source: {{stem meta.file}}
const tokens = {
	{{#each props}}
	{{camelcase name}}: '{{value}}',
	{{/each}}
};
export default tokens;
`
	);
}

**Comment by petekp** [comment]
@mattfelten Nice! Before my current method I was changing the filename from `.js`->`.ts`, running `tsc` and then changing it back to `.js` 😄 

It would be cool to build this into Theo, either as a format or a CLI flag (`--tsDeclarations`).

**Comment by aputinski** [comment]
I'll just chime in and say that I don't have any plans to add TypeScript support, but I'd happily accept a PR for it.

**Comment by petekp** [comment]
I'm down to help implement this. Still wondering what exactly folks would expect from this format. Just definitions? Auto-generated types/interfaces? Thinking optionally outputting `.d.ts` definitions for the built-in JS format would be a reasonable place to start.

**Comment by mattfelten** [comment]
That’s what I would expect. That the output can be packaged and “just work” in typescript projects. I don’t think we would need actual `.ts` files

**Comment by WickyNilliams** [comment]
I used the following which output a TS declaration which matches the `module.js` theo format:

```js
const {
  pipe,
  get,
  map,
  camelCase,
  method,
  sortBy,
  join,
} = require("lodash/fp");

const toConstDeclaration = ({ name, value }) =>
  `export declare const ${camelCase(name)} = ${JSON.stringify(value)}`;

module.exports = pipe(
  method("toJS"),
  get("props"),
  sortBy("name"),
  map(toConstDeclaration),
  join(";\n")
);

```

You can then register this in your setup file:

```
const typescriptFormat = require("./formats/typescript")

module.exports = theo => {
	theo.registerFormat("d.ts", typescriptFormat)
}
